### PR TITLE
[SPARK-9425] [SQL] support DecimalType in UnsafeRow

### PR DESCRIPTION
**Pull request by davies**
This PR brings the support of DecimalType in UnsafeRow, for precision <= 18, it's settable, otherwise it's not settable.

### CODE REVIEW

**Comment by davies** [comment]
ping @rxin

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 1]
```diff
@@ -323,6 +325,22 @@ public double getDouble(int ordinal) {
   }
 
   @Override
+  public Decimal getDecimal(int ordinal, int precision, int scale) {
+    if (isNullAt(ordinal)) {
+      return null;
+    }
+    assertIndexIsValid(ordinal);
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: this should go before the isNullAt check

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala [hunk 2]
```diff
@@ -42,9 +44,64 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
...
+    case BinaryType =>
+      s" + (${ev.isNull} ? 0 : $BinaryWriter.getSize(${ev.primitive}))"
+    case CalendarIntervalType =>
+      s" + (${ev.isNull} ? 0 : 16)"
+    case _: StructType =>
+      s" + (${ev.isNull} ? 0 : $StructWriter.getSize(${ev.primitive}))"
+    case _ => ""
+  }
+
+  def genUpdate(
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: genFieldWriter

**Comment by rxin** [comment]
LGTM

**Comment by rxin** [comment]
Merging this in since the test failure is unrelated.

**Pull request closed without merging**